accessgrant.go

```diff
@@ -0,0 +1,48 @@
+package pubnub
+
+import (
+	"sync"
+
+	"github.com/pubnub/go/messaging"
+)
+
+type AccessGrant struct {
+	pool sync.Pool
+}
+
+type AccessGrantOptions struct {
+	ResumeOnReconnect bool
+	SubscribeTimeout  int
+	Origin            string
+}
+
+func NewAccessGrantOptions() *AccessGrantOptions {
+	return &AccessGrantOptions{
+		ResumeOnReconnect: true,
+		SubscribeTimeout:  20,
+		Origin:            "pubsub.pubnub.com",
+	}
+}
+
+func NewAccessGrant(ao *AccessGrantOptions, cs *ClientSettings) *AccessGrant {
+	messaging.SetResumeOnReconnect(ao.ResumeOnReconnect)
+	messaging.SetSubscribeTimeout(uint16(ao.SubscribeTimeout))
+	messaging.SetOrigin(ao.Origin)
+
+	p := sync.Pool{
+		New: func() interface{} {
+			return NewPubNubClient(cs)
+		},
+	}
+
+	return &AccessGrant{pool: p}
+}
+
+func (ag *AccessGrant) Grant(as *AuthSettings) error {
+	client := ag.pool.Get().(*PubNubClient)
```

Comment:
from the sync.Pool doc `Callers should not assume any relation between values passed to Put and the values returned by Get.` namely better to check if the interface conversion is `ok`